File: python/day14.py
from collections import defaultdict

def step(hist, rules):
    res = defaultdict(int)
    for s, cnt in hist.items():
        if s in rules:
            c = rules[s]
            res[s[0] + c] += cnt
            res[c + s[1]] += cnt
        else:
            res[s] = cnt
    return res

def get_pairs_hist(pattern):
    res = defaultdict(int)
    for i in range(1, len(pattern)):
        res[pattern[i - 1: i + 1]] += 1
    return res

def eval_pattern(start_pattern, rules, steps):
    hist = get_pairs_hist(start_pattern)
    for i in range(steps):
        hist = step(hist, rules)

    char_hist = defaultdict(int)
    for s, cnt in hist.items():
        char_hist[s[0]] += cnt
    char_hist[start_pattern[-1]] += 1

    counts = sorted((char_hist[c], c) for c in char_hist.keys())
    return counts[-1][0] - counts[0][0]

def solution():
    lines = open('../data/14.txt').readlines()
    start_pattern = lines[0].strip()
    rules = {x[0]: x[1] for x in map(lambda s: s.strip().split(" -> "), lines[2:])}

    res1 = eval_pattern(start_pattern, rules, 10)
    res2 = eval_pattern(start_pattern, rules, 40)
    print(f"Answer 1: {res1}\nAnswer 2: {res2}")

File: rust/src/day14.rs
use crate::common;

use std::collections::HashMap;

type RuleSet = HashMap<String, String>;
type PatternMap = HashMap<String, usize>;

fn step(hist: &PatternMap, rules: &RuleSet) -> PatternMap {
    let mut res: PatternMap = HashMap::new();
    for (s, cnt) in hist {
        if rules.contains_key(s) {
            let c = &rules[s];
            *res.entry(s[0..1].to_string() + &c).or_insert(0) += cnt;
            *res.entry(c.to_string() + &s[1..2]).or_insert(0) += cnt;
        } else {
            res.insert(s.to_string(), *cnt);
        }
    }
    res
}

fn get_pairs_hist(pattern: &str) -> PatternMap {
    let mut res = HashMap::new();
    for i in 1..pattern.len() {
        *res.entry(pattern[i - 1..i + 1].to_string()).or_insert(0) += 1;
    }
    res
}

fn eval_pattern(pat: &str, rules: &RuleSet, steps: usize) -> usize {
    let mut hist = get_pairs_hist(&pat);
    for _ in 0..steps {
        hist = step(&hist, &rules);
    }

    let mut char_hist = HashMap::new();
    for (s, cnt) in hist {
        *char_hist.entry(s[0..1].to_string()).or_insert(0) += cnt;
    }
    let last_char = pat[pat.len() - 1..pat.len()].to_string();
    *char_hist.entry(last_char).or_insert(0) += 1;

    let mut counts: Vec<&usize> = char_hist.values().collect();
    counts.sort();
    counts[counts.len() - 1] - counts[0]
}

pub(crate) fn solution() {
    let lines = common::read_lines(&common::data_file(14)).unwrap();
    let start_pattern = lines[0].to_string();
    let rules: RuleSet = lines[2..]
        .iter()
        .map(|s| {
            let parts: Vec<&str> = s.split(" -> ").collect();
            (parts[0].to_string(), parts[1].to_string())
        })
        .collect();
    let res1 = eval_pattern(&start_pattern, &rules, 10);
    let res2 = eval_pattern(&start_pattern, &rules, 40);
    println!("Answer 1: {:?}\nAnswer 2: {:?}", res1, res2);
}
